cpp/arcticdb/codec/typed_block_encoder_impl.hpp

```diff
@@ -0,0 +1,147 @@
+#include <arcticdb/codec/encoding_sizes.hpp>
+#include <arcticdb/codec/passthrough.hpp>
+#include <arcticdb/codec/zstd.hpp>
+#include <arcticdb/codec/lz4.hpp>
+#include <arcticdb/codec/encoded_field.hpp>
+#include <arcticdb/util/buffer.hpp>
+
+#include <type_traits>
+
+namespace arcticdb {
+    /// @todo Split this class. This class does too much. Supports encoding via V1 and V2 but in a clunky way. The
+    ///	interface for encoding is different (V2 requires EncodedBlock to be passed, thus encode_values and
+    /// encode_shapes were added).
+    template<template<typename> class TypedBlock, class TD, EncodingVersion encoder_version>
+    struct TypedBlockEncoderImpl {
+        static size_t max_compressed_size(
+            const arcticdb::proto::encoding::VariantCodec& codec_opts,
+            const TypedBlock<TD>& typed_block
+        ) {
+            return visit_encoder(codec_opts, [&](auto encoder_tag) {
+                return decltype(encoder_tag)::Encoder::max_compressed_size(typed_block);
+            });
+        }
+        /**
+         * Perform encoding of in memory field for storage
+         * @param[in] codec_opts Option used to dispatch to the appropriate encoder and configure it
+         * @param[in] typed_block The block which will be encoded
+         * @param[in, out] field description of the encoding operation
+         * @param[out] out output buffer to write the encoded values to. Must be resized if pos becomes > size
+         * @param[in, out] pos position in bytes in the buffer where to start writing.
+         *  Modified to reflect the position after the last byte written
+         */
+        template <typename EncodedFieldType>
+        static void encode(
+            const arcticdb::proto::encoding::VariantCodec& codec_opts,
+            const TypedBlock<TD>& typed_block,
+            EncodedFieldType& field,
+            Buffer& out,
+            std::ptrdiff_t& pos
+        ) {
+            static_assert(encoder_version == EncodingVersion::V1,
+                "Encoding of both shapes and values at the same time is allowed only in V1 encoding");
+            visit_encoder(codec_opts, [&](auto encoder_tag) {
+                if constexpr(std::is_same_v<typename decltype(encoder_tag)::Encoder, PassthroughEncoder>) {
```

Comment:
Is there a reason why the PassthoughEncoder needs this special treatment?

Reply:
The difference is that there are no encoding options for passthrough encoding. Should I create an empty struct for it?